src/test/java/build/buf/protovalidate/ValidatorDynamicMessageTest.java

```diff
@@ -0,0 +1,138 @@
+        .containsExactly(expectedViolation);
+  }
+
+  @Test
+  public void testOneofConstraintDynamicMessage() throws Exception {
+    DynamicMessage.Builder messageBuilder =
+        createMessageWithUnknownOptions(ExampleOneofConstraints.getDefaultInstance());
+    Violation expectedViolation =
+        Violation.newBuilder()
+            .setFieldPath("contact_info")
+            .setConstraintId("required")
+            .setMessage("exactly one field is required in oneof")
+            .build();
+    assertThat(new Validator().validate(messageBuilder.build()).getViolations())
+        .containsExactly(expectedViolation);
+  }
+
+  @Test
+  public void testMessageConstraintDynamicMessage() throws Exception {
+    DynamicMessage.Builder messageBuilder =
+        createMessageWithUnknownOptions(ExampleMessageConstraints.getDefaultInstance());
+    messageBuilder.setField(
+        messageBuilder.getDescriptorForType().findFieldByName("secondary_email"),
+        "something@somewhere.com");
+    Violation expectedViolation =
+        Violation.newBuilder()
+            .setConstraintId("secondary_email_depends_on_primary")
+            .setMessage("cannot set a secondary email without setting a primary one")
+            .build();
+    assertThat(new Validator().validate(messageBuilder.build()).getViolations())
+        .containsExactly(expectedViolation);
+  }
+
+  private static void gatherDependencies(
+      Descriptors.FileDescriptor fd, Set<DescriptorProtos.FileDescriptorProto> dependencies) {
+    dependencies.add(fd.toProto());
+    for (Descriptors.FileDescriptor dependency : fd.getDependencies()) {
+      gatherDependencies(dependency, dependencies);
+    }
+  }
+
+  private static DescriptorProtos.FileDescriptorSet createFileDescriptorSetForMessage(
+      Descriptors.Descriptor message) {
+    DescriptorProtos.FileDescriptorSet.Builder builder =
+        DescriptorProtos.FileDescriptorSet.newBuilder();
+    Set<DescriptorProtos.FileDescriptorProto> dependencies = new LinkedHashSet<>();
+    gatherDependencies(message.getFile(), dependencies);
+    builder.addAllFile(dependencies);
+    return builder.build();
+  }
+
+  private static Descriptors.FileDescriptor getFileDescriptor(
+      String name, Map<String, DescriptorProtos.FileDescriptorProto> fds)
+      throws Descriptors.DescriptorValidationException {
+    DescriptorProtos.FileDescriptorProto fdProto = fds.get(name);
+    if (fdProto == null) {
+      throw new IllegalArgumentException("unable to file file descriptor proto: " + name);
+    }
+    Descriptors.FileDescriptor[] dependencies =
+        new Descriptors.FileDescriptor[fdProto.getDependencyCount()];
+    for (int i = 0; i < fdProto.getDependencyCount(); i++) {
+      dependencies[i] = getFileDescriptor(fdProto.getDependency(i), fds);
+    }
+    return Descriptors.FileDescriptor.buildFrom(fdProto, dependencies);
+  }
+
+  private static DynamicMessage.Builder createMessageWithUnknownOptions(Message message)
```

Comment:
This is just a round trip operation to go from a message created from Java generated code -> FileDescriptorSet -> DynamicMessage.Builder. The key thing is on line 126-127 where we reparse the FileDescriptorSet from binary format (losing any resolved extensions, etc.).